vvm/utils/versioning.py

```diff
@@ -2,62 +2,61 @@
 import re
 from typing import Any, Optional
 
-from packaging.specifiers import Specifier
+from packaging.specifiers import SpecifierSet
 from packaging.version import Version
 
 from vvm.exceptions import UnexpectedVersionError
 from vvm.install import get_installable_vyper_versions, get_installed_vyper_versions
 
-_VERSION_RE = re.compile(r"\s*#\s*(?:pragma\s+|@)version\s+([=><^~]*)(\d+\.\d+\.\d+\S*)")
+_VERSION_RE = re.compile(r"^(?:#\s*(?:@version|pragma\s+version)\s+(.*))", re.MULTILINE)
 
 
-def _detect_version_specifier(source_code: str) -> Optional[Specifier]:
+def detect_version_specifier_set(source_code: str) -> Optional[SpecifierSet]:
     """
-    Detect the version given by the pragma version in the source code.
+    Detect the specifier set given by the pragma version in the source code.
 
     Arguments
     ---------
     source_code : str
-        Source code to detect the version from.
+        Source code to detect the specifier set from.
 
     Returns
     -------
-    str
-        vyper version specifier, or None if none could be detected.
+    Optional[SpecifierSet]
+        vyper version specifier set, or None if none could be detected.
     """
     match = _VERSION_RE.search(source_code)
     if match is None:
         return None
 
-    specifier, version_str = match.groups()
-    if specifier in ("~", "^"):  # convert from npm-style to pypi-style
-        if Version(version_str) >= Version("0.4.0"):
-            error = "Please use the pypi-style version specifier "
-            error += f"for vyper versions >= 0.4.0 (hint: try ~={version_str})"
-            raise UnexpectedVersionError(error)
-        # for v0.x, both specifiers are equivalent
-        specifier = "~="  # finds compatible versions
+    version_str = match.group(1)
+
+    # X.Y.Z or vX.Y.Z => ==X.Y.Z, ==vX.Y.Z
+    if re.match("[v0-9]", version_str):
+        version_str = "==" + version_str
+    # adapted from vyper/ast/pre_parse.py at commit c32b9b4c6f0d8
+    # convert npm to pep440
+    version_str = re.sub("^\\^", "~=", version_str)
+    version_str = re.sub("^~(?!\\=)", "~=", version_str)
```

Comment:
what does this line do? it's not in the compiler

Reply:
Just removed it, it was doing a `~0.3.0` => `~=0.3.0` conversion, but turns out it doesn't seems to have ever been used in the wild.